check-meta.nix: Add package warnings

AkechiShiro opened the pull request:
## Description of changes
Implements RFC 127 https://github.com/NixOS/rfcs/pull/127
Adds testing for multiple cases

Should supersede : #177272

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AkechiShiro commented:
I will rebase then work on solving conflicts, marking this PR as draft for now.

adisbladis commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -312,17 +379,84 @@ let
-    if metaTypes?${k} then
-      if typeCheck metaTypes.${k} v then
+  # Type of a meta.problems.* value
+  problemTypes = with lib.types; {
```
Module system types are not performant enough. This will have a significant impact on performance.
See https://github.com/NixOS/nixpkgs/pull/273935 for the work I did to bring performance to acceptable levels.

AkechiShiro replied:
I'm working on solving merge conflict son that I can rebase this PR, I saw this part with a comment stating it was better in terms of performance, I will keep the one from upstream.

Thanks for your review and for mentioning this.

AkechiShiro replied:
I believe I kept the change from the PR that you made, marking this as resolved

adisbladis replied:
What I mean is that module system types shouldn't be used at all in meta checks. They're too slow.
This is going to result in a performance regression.

AkechiShiro replied:
Sorry I think I minsunderstood your initial comment, I looked again at your PR and I should avoid using `lib.types` but use some of the types already declared in `meta-types.nix` (this would improve performance)

AkechiShiro replied:
@adisbladis  
I should add to `meta-types.nix` something like this, would this be alright : 
```nix
  77 + │   enum = values: assert isList values && all isString values; {
  78 + │     name = "enum<${concatStringsSep "," values}>";
  79 + │     verify = v: isString v && any (value: v == value) values;
  80 + │   };
```

And add a `let/in` statement for the `problemTypes` to remove the `lib.types`

adisbladis replied:
Yes that looks fine :)
`any (value: v == value) values` could be replaced by `elem v values`

AkechiShiro replied:
Thanks for the help, I think it is ready, I left the old way as a comment but will delete it.

AkechiShiro replied:
It seems `elem` is not defined, should I import it or define it ?

AkechiShiro replied:
I think I found it, it should be all right now.

adisbladis commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -372,7 +444,56 @@ let
+    ++ lib.optionals (meta ? problems) (
+
+      # Check problem kinds are correct
+      lib.pipe meta.problems [
```
This all looks very allocation/iteration heavy:
I'd completely avoid using `lib.pipe` in _any_ hot paths for one.

For a normal code review I wouldn't be so pedantic about this, but check-meta is a performance special case.
Changes here needs to be carefully measured, and the Ofborg performance report doesn't include `checkMeta`, so it's entirely useless in this context.

Measuring nixpkgs eval with `checkMeta = true` before and after this PR (`env NIX_SHOW_STATS=1 nix-env -qa -f ./outpaths.nix --arg checkMeta true > /dev/null`):

- Before
``` json
{
  "cpuTime": 271.2768249511719,
  "envs": {
    "bytes": 7850426888,
    "elements": 413454253,
    "number": 283924554
  },
  "gc": {
    "heapSize": 22804639744,
    "totalBytes": 47139879824
  },
  "list": {
    "bytes": 987373624,
    "concats": 21648163,
    "elements": 123421703
  },
  "nrAvoided": 329081776,
  "nrFunctionCalls": 261979419,
  "nrLookups": 131336801,
  "nrOpUpdateValuesCopied": 675865734,
  "nrOpUpdates": 32046977,
  "nrPrimOpCalls": 146549876,
  "nrThunks": 399878601,
  "sets": {
    "bytes": 15552994336,
    "elements": 911935337,
    "number": 60126809
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 16,
    "Value": 24
  },
  "symbols": {
    "bytes": 2887399,
    "number": 178950
  },
  "values": {
    "bytes": 12350285592,
    "number": 514595233
  }
}
```

- After
``` json
{
  "cpuTime": 296.9483947753906,
  "envs": {
    "bytes": 8506454064,
    "elements": 438912624,
    "number": 312197067
  },
  "gc": {
    "heapSize": 22737526784,
    "totalBytes": 48902439376
  },
  "list": {
    "bytes": 1000276576,
    "concats": 22264431,
    "elements": 125034572
  },
  "nrAvoided": 349063893,
  "nrFunctionCalls": 289412278,
  "nrLookups": 149742381,
  "nrOpUpdateValuesCopied": 694111348,
  "nrOpUpdates": 35108342,
  "nrPrimOpCalls": 168992563,
  "nrThunks": 417292743,
  "sets": {
    "bytes": 15871244064,
    "elements": 927562812,
    "number": 64389942
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 16,
    "Value": 24
  },
  "symbols": {
    "bytes": 4092168,
    "number": 234982
  },
  "values": {
    "bytes": 12774278736,
    "number": 532261614
  }
}
```

AkechiShiro replied:
What could be used to have better performance whilst keeping the same or similar functionality ?

adisbladis replied:
- Don't keep calling mapAttrsToList
  mapAttrsToList is `attrs: map fn (attrNames attrs)`.
  You should store attrNames in a variable and keep reusing that

- lib.remove null -> concatMap
  Nix has an optimisation for empty lists

The code below is a good simple example:

``` nix

# Check that problem has a message (required field)
++ lib.pipe meta.problems [
  (lib.mapAttrsToList (    name: problem:
    if problem ? message then
      null
    else
      "key 'meta.problems.${name}.message' is missing"
    ))
    (lib.remove null)
]
```

->

``` nix
let
  # Stick this in the outermost scope where it makes sense
  problemNames = attrNames meta.problems;
in
  # Nix has an optimisation for returning empty lists
  concatMap (name: if meta.problems.${name} ? message then [] else [ "key 'meta.problems.${name}.message' is missing" ])
```

adisbladis replied:
And just in general: Read the lib implementations you're using here and understand that Nix is not very optimised. 
Sadly we don't have good guidance on how to write performant Nix code.
`NIX_SHOW_STATS=1` is your friend.

adisbladis commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -372,7 +444,56 @@ let
+      # Check that some problem kinds are unqiue
+      ++ lib.pipe meta.problems [
+        (lib.mapAttrs (name: problem: { kind = name; } // problem))
+        # Count the number of instances of each problem kind,
```
I think the manual counting & checking is better implemented by combining `groupBy` & `concatMap` over `problemKindsUnique'`.

qweered commented:
Is there any progress?

infinisil commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -328,6 +392,13 @@ let
+      kind = enum problemKindsManual;
+      message = str;
+      urls = listOf str;
+    };
```
This should be moved, this here would be for adding `meta.problemTypes =` support.

infinisil reviewed (changes requested):
I'm finally getting to review this, the tests are looking pretty good, thanks for making progress on that!

There's some minor issues I've spotted, but other than that I also agree with @adisbladis that there's some work to do regarding performance.

@AkechiShiro I'll work on this a bit and push my changes to this branch, I hope that's okay :)

infinisil committed:
[WIP] stdenv.mkDerivation: RFC 127

Run with `nix-build -A tests.problems`

RFC 127: Add implementation tests

* Also fix merge conflicts
* Make sure that performance PR is conserved
* Use RunCommand instead of alias RunCommandNoCC
* Add enum types to meta-types.nix and use them for problemTypes
* Nixfmt (excepted check-meta.nix)
* "import" elem
* Remove dead code

Co-Authored-By: piegames <git@piegames.de>
Co-Authored-By: AkechiShiro <14914796+AkechiShiro@users.noreply.github.com>
Co-Authored-By: Silvan Mosberger <contact@infinisil.com>

[WIP] Fast problem handler lookup switch

Disable problem check for testing

Better perf

infinisil commented:
I've now optimised this as far as I think is reasonably possible. Performance is still a bit worse than before, but much better now. Check the [automatic performance report](https://github.com/NixOS/nixpkgs/actions/runs/20466544518?pr=338267#summary-58811922543) for details.

I've also adjusted the error messages slightly, while making sure that the tests still pass.

Now the only thing that's left imo are:
- Clean up the code and address some TODOs
- Make sure RFC 127 actually works out well in practice, not sure if that was ever done before

infinisil committed:
More tests and fixes

infinisil commented:
At this point I think it makes sense to open a new PR. Will do so shortly with the latest changes and advertise it for people to try out and give feedback :)
